src/mlpack/methods/ann/layer/celu_impl.hpp

```diff
@@ -0,0 +1,62 @@
+/**
+ * @file celu_impl.hpp
+ * @author Gaurav Singh
+ *
+ * Implementation of the CELU activation function as descibed by Jonathan T. Barron.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_CELU_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_CELU_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "celu.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+CELU<InputDataType, OutputDataType>::CELU(const double alpha) :
+    alpha(alpha),
+    deterministic(false)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void CELU<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, OutputType&& output)
+{
+  Fn(input, output);
+
+  if (!deterministic)
```

Comment:
Where does this come from? I'm not sure I can see that in the paper, maybe I missed something?

Reply:
I think he included it because elu_impl.hpp has this as well.

Reply:
@zoq As @adithya-tp said, the ELU activation function has also been implemented in the same way and from the [paper](https://arxiv.org/pdf/1704.07483.pdf) it can be seen that the author wanted to tackle the problem with ELU which can be seen from the abstract of the paper. So I taught of keeping the template design the same way it is with ELU. If wrong, then I will definitely correct it.

Reply:
Hi @gaurav-singh1998, there is no need for deterministic since it is mentioned in the paper, Sorry I also missed it.

Reply:
@zoq @kartikdutt18 But should deterministic be included because of the derivative being computed only during the time of training and not testing? If wrong, please correct me.

Reply:
> I think he included it because elu_impl.hpp has this as well.

Ahh, I see I think we have to remove that one from the ELU implementation as well.

> @zoq @kartikdutt18 But should deterministic be included because of the derivative being computed only during the time of training and not testing? If wrong, please correct me.

There is no need to distinguish between testing or training, as the calculation doesn't change, one example where it changes is dropout. So let's remove the term here.

Reply:
@zoq So should I correct this and open another pull request correcting the implementation of ELU as well?

Reply:
Yeah, that would be awesome. @ShikharJ is there anything we are missing here, why we want the derivative in the forward function in the elu layer?

Reply:
@zoq Because if we were to see the function 'Deriv' in celu.hpp it can be seen that we need the input data and the function f(x) which is available during the forward pass.

Reply:
@zoq I have incorporated the changes you suggested. Kindly have a look.

Reply:
> @zoq Because if we were to see the function 'Deriv' in 'elu.hpp' it can be seen that we need the input data and the function f(x) which is available during the forward pass? Correct if I am wrong

You are right.

Reply:
Hey, @zoq so is this ready to be merged?

Reply:
So let's reinclude the deterministic parameter, to speed this up only when needed, sorry for the confusion.

Reply:
Hey, @zoq I have the same. So, now is this ready to get merged?

Reply:
Hey, @zoq can you have a look? I have incorporated all the changes you suggested.